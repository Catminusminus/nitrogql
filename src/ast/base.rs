use std::cell::Cell;

use crate::graphql_parser::parser::Rule;
use pest::iterators::Pair;

thread_local! {
    /// Current file to be used when generating Pos.
    static CURRENT_FILE_OF_POS: Cell<usize> = Cell::new(0);
}

pub fn set_current_file_of_pos(file: usize) {
    CURRENT_FILE_OF_POS.with(|cell| cell.set(file));
}

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct Pos {
    /// 0-based line
    pub line: usize,
    /// 0-base column
    pub column: usize,
    /// file (specified by index)
    pub file: usize,
    /// Flag that indicates that this Pos is not from parsed document, but is a built-in structure.
    pub builtin: bool,
}

impl Pos {
    /// Generates a built-in Pos.
    pub fn builtin() -> Self {
        Pos {
            line: 0,
            column: 0,
            file: 0,
            builtin: true,
        }
    }
}

impl From<&Pair<'_, Rule>> for Pos {
    fn from(value: &Pair<'_, Rule>) -> Self {
        let (line, column) = value.line_col();
        // convert 1-based to0-based
        Pos {
            line: line - 1,
            column: column - 1,
            file: CURRENT_FILE_OF_POS.with(|v| v.get()),
            builtin: false,
        }
    }
}

impl Ord for Pos {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.line
            .cmp(&other.line)
            .then(self.column.cmp(&other.column))
    }
}

impl PartialOrd for Pos {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

/// Knows its start position.
pub trait HasPos {
    fn position(&self) -> &Pos;
    fn name(&self) -> Option<&str>;
}

/// Knows its start and content.
pub trait HasSpan {
    fn position(&self) -> &Pos;
    fn name(&self) -> &str;
}

impl<T: HasSpan> HasPos for T {
    fn position(&self) -> &Pos {
        self.position()
    }
    fn name(&self) -> Option<&str> {
        Some(self.name())
    }
}

/// Carrier of name and pos
pub struct NamePos<'a> {
    pub name: Option<&'a str>,
    pub pos: Pos,
}

impl HasPos for NamePos<'_> {
    fn name(&self) -> Option<&str> {
        self.name
    }
    fn position(&self) -> &Pos {
        &self.pos
    }
}

/// Punctuation
#[derive(Copy, Clone, Debug)]
pub struct Punc<'a> {
    pub position: Pos,
    pub token: &'a str,
}

impl HasPos for Punc<'_> {
    fn name(&self) -> Option<&str> {
        None
    }
    fn position(&self) -> &Pos {
        &self.position
    }
}
impl<'a> From<Pair<'a, Rule>> for Punc<'a> {
    fn from(value: Pair<'a, Rule>) -> Self {
        Punc {
            position: (&value).into(),
            token: value.as_str(),
        }
    }
}

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct Keyword<'a> {
    pub name: &'a str,
    pub position: Pos,
}

impl HasPos for Keyword<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        Some(self.name)
    }
}

impl<'a> From<Pair<'a, Rule>> for Keyword<'a> {
    fn from(value: Pair<'a, Rule>) -> Self {
        Keyword {
            position: (&value).into(),
            name: value.as_str(),
        }
    }
}

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct Ident<'a> {
    pub name: &'a str,
    pub position: Pos,
}

impl HasPos for Ident<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        Some(self.name)
    }
}

impl<'a> From<Pair<'a, Rule>> for Ident<'a> {
    fn from(value: Pair<'a, Rule>) -> Self {
        Ident {
            position: (&value).into(),
            name: value.as_str(),
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct Variable<'a> {
    /// Variable name that does not include '$'
    pub name: &'a str,
    /// Position of '$'
    pub position: Pos,
}

impl HasPos for Variable<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        Some(self.name)
    }
}
