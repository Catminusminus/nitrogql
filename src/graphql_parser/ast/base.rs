use super::super::parser::Rule;
use pest::iterators::Pair;

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct Pos {
    /// 0-based line
    pub line: usize,
    /// 0-base column
    pub column: usize,
    /// Flag that indicates that this Pos is not from parsed document, but is a built-in structure.
    pub builtin: bool,
}

impl Pos {
    /// Generates a built-in Pos.
    pub fn builtin() -> Self {
        Pos {
            line: 0,
            column: 0,
            builtin: true,
        }
    }
}

impl From<&Pair<'_, Rule>> for Pos {
    fn from(value: &Pair<'_, Rule>) -> Self {
        let (line, column) = value.line_col();
        // convert 1-based to0-based
        Pos {
            line: line - 1,
            column: column - 1,
            builtin: false,
        }
    }
}

/// Knows its start position.
pub trait HasPos {
    fn position(&self) -> &Pos;
    fn name(&self) -> Option<&str>;
}

/// Knows its start and end position.
pub trait HasSpan {
    fn span(&self) -> (&Pos, &Pos);
    fn name(&self) -> Option<&str>;
}

impl<T: HasSpan> HasPos for T {
    fn position(&self) -> &Pos {
        self.span().0
    }
    fn name(&self) -> Option<&str> {
        self.name()
    }
}

/// Carrier of name and pos
pub struct NamePos<'a> {
    pub name: Option<&'a str>,
    pub pos: Pos,
}

impl HasPos for NamePos<'_> {
    fn name(&self) -> Option<&str> {
        self.name
    }
    fn position(&self) -> &Pos {
        &self.pos
    }
}

/// Punctuation
#[derive(Copy, Clone, Debug)]
pub struct Punc<'a> {
    pub position: Pos,
    pub token: &'a str,
}

impl HasPos for Punc<'_> {
    fn name(&self) -> Option<&str> {
        None
    }
    fn position(&self) -> &Pos {
        &self.position
    }
}

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct Ident<'a> {
    pub name: &'a str,
    pub position: Pos,
}

impl HasPos for Ident<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        Some(self.name)
    }
}

impl<'a> From<Pair<'a, Rule>> for Ident<'a> {
    fn from(value: Pair<'a, Rule>) -> Self {
        Ident {
            position: (&value).into(),
            name: value.as_str(),
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct Variable<'a> {
    /// Variable name that does not include '$'
    pub name: &'a str,
    /// Position of '$'
    pub position: Pos,
}

impl HasPos for Variable<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        Some(self.name)
    }
}
