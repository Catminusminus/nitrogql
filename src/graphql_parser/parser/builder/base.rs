use super::Rule;
use pest::iterators::Pair;

use crate::graphql_parser::ast::base::Variable;

pub fn build_variable(pair: Pair<Rule>) -> Variable {
    Variable {
        name: pair.as_str(),
        position: (&pair).into(),
    }
}
