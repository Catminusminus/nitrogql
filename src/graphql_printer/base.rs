use json_writer::JSONWriterValue;

use crate::{
    graphql_parser::ast::{
        base::Ident,
        r#type::Type,
        value::{StringValue, Value},
    },
    source_map_writer::writer::SourceMapWriter,
};

use super::GraphQLPrinter;

impl GraphQLPrinter for Type<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        match self {
            Type::Named(name) => {
                writer.write_for(&name.name.name, self);
            }
            Type::NonNull(non_null) => {
                non_null.r#type.print_graphql(writer);
                writer.write("!");
            }
            Type::List(list) => {
                writer.write("[");
                list.r#type.print_graphql(writer);
                writer.write("]");
            }
        }
    }
}

impl GraphQLPrinter for Value<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        match self {
            Value::BooleanValue(value) => {
                writer.write(value.keyword);
            }
            Value::IntValue(value) => {
                writer.write(value.value);
            }
            Value::FloatValue(value) => {
                writer.write(value.value);
            }
            Value::StringValue(value) => {
                value.print_graphql(writer);
            }
            Value::EnumValue(value) => {
                writer.write_for(value.value, self);
            }
            Value::NullValue(value) => {
                writer.write(value.keyword);
            }
            Value::Variable(value) => {
                value.print_graphql(writer);
            }
            Value::ListValue(value) => {
                writer.write("[");
                for (idx, v) in value.values.iter().enumerate() {
                    if idx > 0 {
                        writer.write(",");
                    }
                    v.print_graphql(writer);
                }
                writer.write("]");
            }
            Value::ObjectValue(value) => {
                writer.write("{");
                if value.fields.len() < 2 {
                    for (name, value) in value.fields.iter() {
                        name.print_graphql(writer);
                        writer.write(": ");
                        value.print_graphql(writer);
                    }
                } else {
                    writer.write("\n");
                    writer.indent();
                    for (name, value) in value.fields.iter() {
                        name.print_graphql(writer);
                        writer.write(": ");
                        value.print_graphql(writer);
                        writer.write("\n");
                    }
                    writer.dedent();
                }
                writer.write("}");
            }
        }
    }
}

impl GraphQLPrinter for StringValue {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        let mut result = String::with_capacity(self.value.capacity());
        let is_multiline = self.value.find('\n').is_some();
        if is_multiline {
            // print as multiline string
            result.push_str("\"\"\"");
            let mut dq_count: usize = 0;
            for c in self.value.chars() {
                if c != '"' {
                    if dq_count > 0 {
                        result.push_str(&"\"".repeat(dq_count));
                        dq_count = 0;
                    }
                    result.push(c);
                    continue;
                }
                dq_count += 1;
                if dq_count == 3 {
                    // """ in string
                    result.push_str("\\\"\"\"");
                    dq_count = 0;
                }
            }
            if dq_count > 0 {
                result.push_str(&"\"".repeat(dq_count));
                dq_count = 0;
            }
            result.push_str("\"\"\"");
            writer.write(&result);
        } else {
            // single line string
            result.push('"');
            for c in self.value.chars() {
                match c {
                    '\r' => result.push_str("\\r"),
                    '\n' => result.push_str("\\n"),
                    c if c.is_control() => {
                        result.push_str(&format!("\\u{{{:x}}}", c as u32));
                    }
                    c => result.push(c),
                }
            }
            result.push('"');
            writer.write(&result);
        }
    }
}

impl GraphQLPrinter for Ident<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        writer.write_for(self.name, self);
    }
}
