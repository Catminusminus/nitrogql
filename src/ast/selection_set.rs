use super::{
    base::{HasPos, Ident, Pos},
    directive::Directive,
    value::Arguments,
};

#[derive(Clone, Debug)]
pub struct SelectionSet<'a> {
    pub position: Pos,
    pub selections: Vec<Selection<'a>>,
}

impl HasPos for SelectionSet<'_> {
    fn position(&self) -> &Pos {
        &self.position
    }
    fn name(&self) -> Option<&str> {
        None
    }
}

#[derive(Clone, Debug)]
pub enum Selection<'a> {
    Field(Field<'a>),
    FragmentSpread(FragmentSpread<'a>),
    InlineFragment(InlineFragment<'a>),
}

#[derive(Clone, Debug)]
pub struct Field<'a> {
    pub alias: Option<Ident<'a>>,
    pub name: Ident<'a>,
    pub arguments: Option<Arguments<'a>>,
    pub directives: Vec<Directive<'a>>,
    pub selection_set: Option<SelectionSet<'a>>,
}

#[derive(Clone, Debug)]
pub struct FragmentSpread<'a> {
    pub position: Pos,
    pub fragment_name: Ident<'a>,
    pub directives: Vec<Directive<'a>>,
}

#[derive(Clone, Debug)]
pub struct InlineFragment<'a> {
    pub position: Pos,
    pub type_condition: Option<Ident<'a>>,
    pub directives: Vec<Directive<'a>>,
    pub selection_set: SelectionSet<'a>,
}
