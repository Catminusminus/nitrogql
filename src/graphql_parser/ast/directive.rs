use super::{
    base::{Ident, Pos},
    value::Arguments,
};

#[derive(Clone, Debug)]
pub struct Directive<'a> {
    pub position: Pos,
    /// Name of directive (does not include '@')
    pub name: Ident<'a>,
    pub arguments: Option<Arguments<'a>>,
}
