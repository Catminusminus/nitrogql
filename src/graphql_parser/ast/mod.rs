#[derive(Clone, Debug)]
pub struct OperationDocument<'a> {
    pub definitions: Vec<OperationDefinition<'a>>,
}

#[derive(Clone, Debug)]
pub struct OperationDefinition<'a> {
    pub source: &'a str,
}
