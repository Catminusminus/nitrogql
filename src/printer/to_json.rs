use graphql_parser::{
    query::{
        Definition, Document, Field, FragmentSpread, InlineFragment, OperationDefinition,
        Selection, SelectionSet, TypeCondition, VariableDefinition,
    },
    schema::{Directive, Text, Type, Value},
};
use json_writer::JSONObjectWriter;

use super::helpers::{Argument, JSONValue, Name, Variable};

/// Value that can be printed into JSON.
pub trait JsonPrintable {
    fn print_json(&self, writer: &mut JSONObjectWriter);
}

impl JsonPrintable for Document<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "Document");

        let mut definitions_writer = writer.array("definitions");
        for d in &self.definitions {
            match d {
                Definition::Operation(op) => {
                    op.print_json(&mut definitions_writer.object());
                }
                _ => {
                    todo!("operation definition")
                }
            }
        }
    }
}

impl JsonPrintable for OperationDefinition<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "OperationDefinition");
        match self {
            OperationDefinition::Query(query) => {
                writer.value("operation", "query");
                if let Some(name) = &query.name {
                    writer.value("name", JSONValue(&Name(name.as_str())));
                }
                let mut variable_definitions_writer = writer.array("variableDefinitions");
                for v in &query.variable_definitions {
                    v.print_json(&mut variable_definitions_writer.object());
                }
                variable_definitions_writer.end();
                let mut directives_writer = writer.array("directives");
                for v in &query.directives {
                    v.print_json(&mut directives_writer.object());
                }
                directives_writer.end();
                write_selection_set(&query.selection_set, writer);
            }
            _ => {}
        }
    }
}

impl JsonPrintable for VariableDefinition<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "VariableDefinition");
        writer.value(
            "variable",
            JSONValue(&Variable::new(Name(self.name.as_str()))),
        );
        writer.value("type", JSONValue(&self.var_type));
        todo!("More fields to add")
    }
}

impl JsonPrintable for Directive<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "Directive");
        writer.value("name", JSONValue(&Name(self.name.as_str())));
        let mut arguments_writer = writer.array("arguments");
        for (name, value) in &self.arguments {
            Argument::new(name, value).print_json(&mut arguments_writer.object());
        }
    }
}

impl<'a, T> JsonPrintable for Type<'a, T>
where
    T: Text<'a>,
{
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        match self {
            Type::NonNullType(inner) => {
                writer.value("kind", "NonNullType");
                inner.print_json(&mut writer.object("type"));
            }
            Type::NamedType(inner) => {
                writer.value("kind", "NamedType");
                writer.value("name", JSONValue(&Name(inner.as_ref())));
            }
            Type::ListType(inner) => {
                writer.value("kind", "ListType");
                inner.print_json(&mut writer.object("type"));
            }
        }
    }
}

impl JsonPrintable for Value<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        match self {
            Value::Variable(v) => {
                Variable::new(Name(v)).print_json(writer);
            }
            Value::Boolean(b) => {
                writer.value("kind", "BooleanValue");
                writer.value("value", *b);
            }
            Value::Int(i) => {
                writer.value("kind", "IntValue");
                writer.value("value", &i.as_i64().unwrap().to_string());
            }
            Value::Float(f) => {
                writer.value("kind", "FloatValue");
                writer.value("value", &f.to_string());
            }
            Value::String(s) => {
                writer.value("kind", "StringValue");
                writer.value("value", s);
            }
            Value::Null => {
                writer.value("kind", "NulLValue");
            }
            Value::List(list) => {
                writer.value("kind", "ListValue");
                let mut values_writer = writer.array("values");
                for v in list {
                    v.print_json(&mut values_writer.object());
                }
            }
            Value::Object(obj) => {
                writer.value("kind", "ObjectValue");
                let mut fields_writer = writer.array("fields");
                for (key, value) in obj {
                    let mut field_writer = fields_writer.object();
                    field_writer.value("kind", "ObjectField");
                    field_writer.value("name", JSONValue(&Name(key)));
                    value.print_json(&mut field_writer.object("value"));
                }
            }
            Value::Enum(e) => {
                writer.value("kind", "EnumValue");
                writer.value("value", e);
            }
        }
    }
}

impl JsonPrintable for SelectionSet<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "SelectionSet");
        let mut selections_writer = writer.array("selections");
        for selection in self.items.iter() {
            selection.print_json(&mut selections_writer.object());
        }
    }
}

impl JsonPrintable for Selection<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        match self {
            Selection::Field(field) => {
                field.print_json(writer);
            }
            Selection::FragmentSpread(frag) => {
                frag.print_json(writer);
            }
            Selection::InlineFragment(frag) => {
                frag.print_json(writer);
            }
        }
    }
}

impl JsonPrintable for Field<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "Field");
        writer.value("name", JSONValue(&Name(&self.name)));
        if let Some(ref alias) = self.alias {
            writer.value("alias", JSONValue(&Name(alias)));
        }

        let mut arguments_writer = writer.array("arguments");
        for (name, value) in self.arguments.iter() {
            Argument::new(name, value).print_json(&mut arguments_writer.object());
        }
        arguments_writer.end();
        let mut directives_writer = writer.array("directives");
        for d in self.directives.iter() {
            d.print_json(&mut directives_writer.object());
        }
        directives_writer.end();
        write_selection_set(&self.selection_set, writer);
    }
}

impl JsonPrintable for FragmentSpread<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "FragmentSpread");
        writer.value("name", JSONValue(&Name(&self.fragment_name)));
        let mut directives_writer = writer.array("directives");
        for d in self.directives.iter() {
            d.print_json(&mut directives_writer.object());
        }
    }
}

impl JsonPrintable for InlineFragment<'_, String> {
    fn print_json(&self, writer: &mut JSONObjectWriter) {
        writer.value("kind", "InlineFragment");
        if let Some(ref cond) = self.type_condition {
            match cond {
                TypeCondition::On(ref ty) => {
                    let mut type_condition_writer = writer.object("typeCondition");
                    Type::NamedType::<&str>(ty).print_json(&mut type_condition_writer);
                }
            }
        }
        let mut directives_writer = writer.array("directives");
        for d in self.directives.iter() {
            d.print_json(&mut directives_writer.object());
        }
        directives_writer.end();
        write_selection_set(&self.selection_set, writer);
    }
}

fn write_selection_set(set: &SelectionSet<'_, String>, writer: &mut JSONObjectWriter) {
    if set.items.is_empty() {
        return;
    }
    let mut selection_set_writer = writer.object("selectionSet");
    set.print_json(&mut selection_set_writer);
}
