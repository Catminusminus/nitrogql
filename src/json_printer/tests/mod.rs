#[cfg(test)]
mod tests {
    use json_writer::JSONObjectWriter;

    use crate::graphql_parser::ast::base::Pos;

    use super::super::to_json::JsonPrintable;

    static POS: Pos = Pos { line: 0, column: 0 };

    mod value {
        use super::{print_json_to_string, POS};
        use crate::graphql_parser::ast::{
            base::Ident,
            value::{
                BooleanValue, EnumValue, FloatValue, IntValue, ListValue, NullValue, ObjectValue,
                StringValue, Value,
            },
        };
        use insta::assert_snapshot;

        #[test]
        fn null_value() {
            assert_snapshot!(print_json_to_string(Value::NullValue(NullValue {
                position: POS,
                keyword: "null"
            })))
        }
        #[test]
        fn boolean_value() {
            assert_snapshot!(print_json_to_string(Value::BooleanValue(BooleanValue {
                position: POS,
                keyword: "false",
                value: false,
            })));
            assert_snapshot!(print_json_to_string(Value::BooleanValue(BooleanValue {
                position: POS,
                keyword: "true",
                value: true,
            })));
        }
        #[test]
        fn int_value() {
            assert_snapshot!(print_json_to_string(Value::IntValue(IntValue {
                position: POS,
                value: "123"
            })));
            assert_snapshot!(print_json_to_string(Value::IntValue(IntValue {
                position: POS,
                value: "-5"
            })));
        }
        #[test]
        fn float_value() {
            assert_snapshot!(print_json_to_string(Value::FloatValue(FloatValue {
                position: POS,
                value: "123.4"
            })));
            assert_snapshot!(print_json_to_string(Value::FloatValue(FloatValue {
                position: POS,
                value: "-0.05"
            })));
        }
        #[test]
        fn string_value() {
            assert_snapshot!(print_json_to_string(Value::StringValue(StringValue {
                position: POS,
                value: ""
            })));
            assert_snapshot!(print_json_to_string(Value::StringValue(StringValue {
                position: POS,
                value: "foobar"
            })));
        }
        #[test]
        fn enum_value() {
            assert_snapshot!(print_json_to_string(Value::EnumValue(EnumValue {
                position: POS,
                value: "A"
            })))
        }
        #[test]
        fn list_value() {
            assert_snapshot!(print_json_to_string(Value::ListValue(ListValue {
                position: POS,
                values: vec![
                    Value::IntValue(IntValue {
                        position: POS,
                        value: "1"
                    }),
                    Value::IntValue(IntValue {
                        position: POS,
                        value: "2"
                    }),
                    Value::IntValue(IntValue {
                        position: POS,
                        value: "3"
                    }),
                ]
            })))
        }
        #[test]
        fn object_value() {
            assert_snapshot!(print_json_to_string(Value::ObjectValue(ObjectValue {
                position: POS,
                fields: vec![(
                    Ident {
                        position: POS,
                        name: "foo"
                    },
                    Value::NullValue(NullValue {
                        position: POS,
                        keyword: "null"
                    })
                )]
            })));
        }
    }

    mod r#type {
        use insta::assert_snapshot;

        use crate::{
            graphql_parser::ast::{
                base::Ident,
                r#type::{ListType, NamedType, NonNullType, Type},
            },
            json_printer::tests::tests::{print_json_to_string, POS},
        };

        #[test]
        fn named_type() {
            assert_snapshot!(print_json_to_string(Type::Named(NamedType {
                name: Ident {
                    position: POS,
                    name: "tyty"
                }
            })));
        }

        #[test]
        fn non_null_type() {
            assert_snapshot!(print_json_to_string(Type::NonNull(Box::new(NonNullType {
                r#type: Type::Named(NamedType {
                    name: Ident {
                        position: POS,
                        name: "tyty"
                    }
                })
            }))));
        }

        #[test]
        fn list_type() {
            assert_snapshot!(print_json_to_string(Type::List(Box::new(ListType {
                position: POS,
                r#type: Type::Named(NamedType {
                    name: Ident {
                        position: POS,
                        name: "String"
                    }
                })
            }))));
        }
    }

    fn print_json_to_string<V>(value: V) -> String
    where
        V: JsonPrintable,
    {
        let mut result = String::new();
        let mut writer = JSONObjectWriter::new(&mut result);
        value.print_json(&mut writer);
        writer.end();
        result
    }
}
