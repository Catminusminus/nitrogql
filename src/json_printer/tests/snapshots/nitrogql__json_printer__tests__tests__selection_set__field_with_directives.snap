---
source: src/json_printer/tests/mod.rs
expression: "print_json_to_string(Field {\n        alias: None,\n        name: Ident { position: POS, name: \"field\" },\n        arguments: None,\n        directives: vec![Directive\n            {\n                position : POS, name : Ident\n                { name : \"dedede\", position : POS }, arguments : None\n            }, Directive\n            {\n                position : POS, name : Ident\n                { name : \"dedede2\", position : POS }, arguments :\n                Some(Arguments\n                {\n                    position : POS, arguments : vec!\n                    [(Ident { name : \"limit\", position : POS }, Value ::\n                    IntValue(IntValue { position : POS, value : \"10\" }))]\n                })\n            }],\n        selection_set: None,\n    })"
---
{"kind":"Field","name":{"type":"Name","value":"field"},"arguments":[],"directives":[{"kind":"Directive","name":{"type":"Name","value":"dedede"},"arguments":[]},{"kind":"Directive","name":{"type":"Name","value":"dedede2"},"arguments":[{"kind":"Argument","name":{"type":"Name","value":"limit"},"value":{"kind":"IntValue","value":"10"}}]}]}
